R/subsampleCounts.R

```diff
@@ -0,0 +1,257 @@
+#' Microbiome Datasets Are Compositional: And This Is Not Optional.
+#' Frontiers in Microbiology 8: 2224. doi: 10.3389/fmicb.2017.02224
+#' 
+#' Weiss S, Xu ZZ, Peddada S, Amir A, Bittinger K, Gonzalez A, Lozupone C, 
+#' Zaneveld JR, Vázquez-Baeza Y, Birmingham A, Hyde ER. Normalization and 
+#' microbial differential abundance strategies depend upon data characteristics. 
+#' Microbiome. 2017 Dec;5(1):1-8.
+#' 
+#' @return Either a \code{TreeSummarizedExperiment} or a 
+#'   \code{MultiAssayExperiment} object
+#' 
+#' @author Sudarshan A. Shetty 
+#' 
+#' @name subsampleCounts
+#'  
+#' @examples
+#' # When samples in TreeSE are less than specified min_size, they will be removed.
+#' # If after subsampling features are not present in any of the samples, 
+#' # they will be removed.
+#' data("GlobalPatterns")
+#' tse <- GlobalPatterns
+#' tse.subsampled <- subsampleCounts(tse, 
+#'                                   min_size = 60000, 
+#'                                   name = "subsampled", 
+#'                                   seed = 123,
+#'                                   return_type = "TreeSummarizedExperiment")
+#' tse.subsampled
+#' dim(tse)
+#' dim(tse.subsampled)
+#' 
+#' # Return MultiAssayExperiment
+#' mae <- subsampleCounts(tse, 
+#'                        min_size = 60000, 
+#'                        name = "subsampled",
+#'                        seed = 123,
+#'                        return_type = "MultiAssayExperiment")
+#' mae
+#' 
+NULL
+
+#' @rdname subsampleCounts
+#' @aliases rarifyCounts
+#' @export
+setGeneric("subsampleCounts", signature = c("x"),
+  function(x, abund_values = "counts", min_size = min(colSums2(assay(x))),
+           seed = runif(1, 0, .Machine$integer.max), replace = TRUE,
+           name = "subsampled", verbose = TRUE, 
+           return_type=c("TreeSummarizedExperiment", "MultiAssayExperiment"), ...)
+    standardGeneric("subsampleCounts"))
+
+#' @importFrom SummarizedExperiment assay assay<-
+#' @importFrom DelayedMatrixStats colSums2 rowSums2
+#' @rdname subsampleCounts
+#' @aliases rarifyCounts
+#' @export
+setMethod("subsampleCounts", signature = c(x = "SummarizedExperiment"),
+  function(x, abund_values = "counts", min_size = min(colSums2(assay(x))),
+           seed = runif(1, 0, .Machine$integer.max), replace = TRUE, 
+           name = "subsampled", verbose = TRUE, 
+           return_type=c("TreeSummarizedExperiment", "MultiAssayExperiment"), ...){
+    warning("Subsampling/Rarefying may undermine downstream analyses",
+            "\nand have unintended consequences. Therefore, make sure",
+            "\nthis normalization is appropriate for your data. \n ",
+            call. = FALSE)
+    # Input check
+    .check_assay_present(abund_values, x)
+    
+    # check return_type
+    if(!return_type[1] %in% c("TreeSummarizedExperiment", "MultiAssayExperiment")){
+      message("`return_type` must be either `TreeSummarizedExperiment`",
+              "or `MultiAssayExperiment`")
+    }
```

Comment:
Usually you check this via `match.arg`. I don't see a point why it shouldn't apply here.

Side question: What is the difference between in return type? If it is a conveniance thing, remove it.

Reply:
#112 here we discussed this and the consensus was to either provide the `TSE` (which is the input class) or modified `TSE` within an `MAE` because rownames and colnames can differ after subsampling. I am unable to understand the rationale for returning to a different class like `SCE`. If input `TSE` has a tree, will SCE allow `rowTree` and `referenceSeq` slots?

Reply:
> ... the `TSE` (which is the input class) 

does it need to be? You dispatch to `SE` in line 112 so if that works, why not return the same class as the input?

If you want to have different results, based on the input class, you should the specific methods. Possibly they each wrap each other.

Back to the current state: If it is not necessary to have an MAE, than don't return it. The users can convert the object themselves. However, if you need MAE functionality always, than return always an MAE. If you need it in some cases, return it those cases and document those cases in  the man page.

But the user doesn't need to choose the output type via an argument. That can be done like this as mentioned above:

```r
subsamples_se <- subsampleCounts(se)
mae <- MultiAssayExperiment::MultiAssayExperiment(c("original" = se,"subsampled" = subsamples_se ))
```
Here the user chooses actively via the normal constructor.

Reply:
Did I talk in circles on the issue thread? Sorry for that

Reply:
> I will simplify this function then it is a straightforward implementation without the need for `altExp` or `MAE`, just a `modified-TSE` . I can change `signature = c(x = "SummarizedExperiment")` to `signature = c(x = "TreeSummarizedExperiment")`

I would keep the former. You are not subsampling the tree are you? So technically there is need for a TSE as input

Reply:
True. I will keep `SummarizedExperiment`